### PR TITLE
[controller] Network commissioning support on Controller

### LINKED ISSUE #3945: Add network provisioning cluster to CHIP
#### Problem
per spec https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/309, we need initial network provisioning cluster xml

### PR DESCRIPTION
**Pull request by erjiaqing**
#### Problem
Add network provisioning cluster define for provisioning use, also adds support on controller side.

*Support on end device side will be a seperate PR later.*

This PR also fixs support for ByteString in encoder for interaction model.

 #### Summary of Changes
See above.

 Fixes #3945

### CODE REVIEW

**Review by andy31415 — APPROVED** [review]

File: src/app/encoder.cpp [hunk 1]
```diff
@@ -2467,6 +2479,259 @@ PacketBufferHandle encodeLevelControlClusterReadClusterRevisionAttribute(Endpoin
...
+    if (!CanCastTo<uint8_t>(operationalDatasetStrLen))
+    {
+        ChipLogError(Zcl, "Error encoding %s command. String too long: %d", kName, operationalDatasetStrLen);
+        return PacketBufferHandle();
+    }
+    buf.Put8(kFrameControlClusterSpecificCommand)
+        .Put8(kSeqNum)
+        .Put8(ZCL_ADD_THREAD_NETWORK_COMMAND_ID)
+        .Put(static_cast<uint8_t>(operationalDatasetStrLen))
+        .Put(reinterpret_cast<char *>(operationalDataset))
```
Comment (LuDuda) [inline src/app/encoder.cpp:2518]: Just a question.. Why we send Operational Dataset in a form of a string and not in the binary form of Operational Dataset TLV?

**Review by LuDuda — APPROVED** [review]
@erjiaqing thank you for doing this! LGTM 👍 

Few questions - more about specification..:
- How controller obtains Thread Operational Credentials (is it out of the spec, and we should use MeshCop connection to Thread Border Router)?
- How Controller will know if the device operates on WiFi or Thread? Or the user will always need to know that and configure it manually?

File: src/app/encoder.cpp (hunk 1 above)
Comment (erjiaqing) [reply src/app/encoder.cpp:2518]: @vivien-apple I think we need BYTE_STRING type support in encoder and dispatcher. (item 1 in #4503 )

**Review by yunhanw-google — APPROVED** [review]

File: src/app/zap-templates/templates/chip/CHIPClusters-src.zapt [hunk 2]
```diff
@@ -15,7 +15,7 @@ namespace Controller {
 
 // {{asCamelCased name false}} Cluster Commands
 {{#chip_server_cluster_commands}}
-CHIP_ERROR {{asCamelCased clusterName false}}Cluster::{{asCamelCased name false}}(Callback::Callback<> * onCompletion{{#chip_server_cluster_command_arguments}}, {{chipType}} {{asCamelCased label}}{{/chip_server_cluster_command_arguments}})
+CHIP_ERROR {{asCamelCased clusterName false}}Cluster::{{asCamelCased name false}}(Callback::Callback<> * onCompletion{{#chip_server_cluster_command_arguments}}, {{chipType}} {{asCamelCased label}}{{#if (isByteString type)}}, uint32_t {{asCamelCased label}}Len{{/if}}{{/chip_server_cluster_command_arguments}})
```
Comment (vivien-apple) [inline src/app/zap-templates/templates/chip/CHIPClusters-src.zapt:18]: We can do better here. The whole point to `chip_server_cluster_command_arguments` was to try to minimise the amount of additional code in the template.

It is OK to do  {{#if ...}}...{{/if}}` blocks inside for some exceptional cases that should not be replicated all over the uses of `chip_server_cluster_command_arguments` but for this case what we should do instead if to fix the `chip_server_cluster_command_arguments` function code in `helper.js` to generate 2 arguments for `BYTE_STRING` instead of 1. Then it will be transparent for all consumers of the API.

File: src/app/zap-templates/zcl/chip-devices.xml [hunk 3]
```diff
@@ -186,6 +186,8 @@ limitations under the License.
       </include>
       <include cluster="Temperature Measurement" client="false" server="true" clientLocked="true" serverLocked="true">
       </include>
+      <include cluster="Network Provisioning" client="true" server="true" clientLocked="true" serverLocked="true">
```
Comment (vivien-apple) [inline src/app/zap-templates/zcl/chip-devices.xml:189]: I think we don't want the client here but that's a matter of which patch lands first between this one and #4661... One of us will have to rebase some work :)

File: src/app/zap-templates/zcl/provisioning.xml [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+        <item name="OtherConnectionFailure" value="0x9"/>
+        <item name="IPV6Failed" value="0xa"/>
+        <item name="IPBindFailed" value="0xb"/>
+        <item name="Label9" value="0xc"/>
+        <item name="Label10" value="0xd"/>
+        <item name="Label11" value="0xe"/>
+        <item name="Label12" value="0xf"/>
+        <item name="Label13" value="0x10"/>
+        <item name="Label14" value="0x11"/>
+        <item name="Label15" value="0x12"/>
```
Comment (vivien-apple) [inline src/app/zap-templates/zcl/provisioning.xml:38]: Do we really want those `Label**` strings here ?

File: examples/all-clusters-app/all-clusters-common/all-clusters-app.zap [hunk 5]
```diff
@@ -5667,6 +5667,196 @@
...
+        {
+          "name": "Network Provisioning",
+          "code": 43690,
+          "mfgCode": null,
+          "define": "NWPROV_CLUSTER",
+          "side": "server",
+          "enabled": 1,
+          "commands": [
+            {
+              "name": "ScanNetworksResp",
```
Comment (vivien-apple) [inline examples/all-clusters-app/all-clusters-common/all-clusters-app.zap:5779]: Please use `Response` and not `Resp` for all responses.

File: src/app/zap-templates/zcl/provisioning.xml [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+            <arg name="Credentials" type="OCTET_STRING"/>
+            <arg name="Breadcrumb" type="INT64U"/>
+            <arg name="TimeoutMs" type="INT32U"/>
+        </command>
+        <command source="server" code="0x03" name="AddWiFiNetworkResp" optional="false" cli="chip nwprov addwifiresponse">
+            <description>TODO</description>
+            <arg name="ErrorCode" type="INT8U"/>
+            <arg name="DebugText" type="CHAR_STRING"/>
+        </command>
+        <command source="client" code="0x04" name="UpdateWiFiNetwork" optional="false" cli="chip nwprov updatewifinetwork">
```
Comment (vivien-apple) [inline src/app/zap-templates/zcl/provisioning.xml:91]: I guess the `WiFi` bits could be `optional="true"` ?

File: src/app/zap-templates/zcl/provisioning.xml [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+        <domain>CHIP</domain>
+        <description>TODO</description>
+        <code>0xaaaa</code>
+        <define>NWPROV_CLUSTER</define>
+        <client tick="false" init="false">true</client>
+        <server tick="false" init="false">true</server>
+        <command source="client" code="0x00" name="ScanNetworks" optional="false" cli="chip nwprov scannetworks">
+            <description>TODO</description>
+            <arg name="SSID" type="OCTET_STRING"/>
+            <arg name="Breadcrumb" type="INT64U"/>
```
Comment (vivien-apple) [inline src/app/zap-templates/zcl/provisioning.xml:69]: AFAICT most of the current clusters definitions use `camelCase` for the name of the arguments. Let's stay consistent.

**Review by vivien-apple — APPROVED** [review]

**Review by vivien-apple — CHANGES_REQUESTED** [review]
Sorry I meant to request changes.

**Review by mspang — COMMENTED** [review]
I think we should do 

nwprov -> network[-_]provisioning

throughout the patch, including CLI.

File: src/app/zap-templates/zcl/provisioning.xml (hunk 4 above)
Comment (erjiaqing) [reply src/app/zap-templates/zcl/provisioning.xml:38]: Just follows the spec, we will update this file as the spec is updating.

File: src/app/zap-templates/zcl/chip-devices.xml (hunk 3 above)
Comment (erjiaqing) [reply src/app/zap-templates/zcl/chip-devices.xml:189]: The network provisioning cluster on client will accepts *Response commands, I have no idea what is the meaning of "client="true"", "server="true"" here.

File: src/app/zap-templates/templates/chip/CHIPClusters-src.zapt (hunk 2 above)
Comment (erjiaqing) [reply src/app/zap-templates/templates/chip/CHIPClusters-src.zapt]: Let's trace this in #4503

**Comment by erjiaqing** [comment]
@vivien-apple PTAL

**Review by mspang — APPROVED** [review]

File: src/app/zap-templates/zcl/chip-devices.xml (hunk 3 above)
Comment (vivien-apple) [reply src/app/zap-templates/zcl/chip-devices.xml:189]: > The network provisioning cluster on client will accepts *Response commands, I have no idea what is the meaning of "client="true"", "server="true"" here.

My understanding is that those fields are to enable/disable a cluster by default for a particular device configuration. I just disable all client clusters from the `all-clusters-app.zap` file. I did a quick look at this file and most `client` are true by default.

So I guess it does not matter too much what the value of the field in this file is for this patch. We will need to take a look at that in a different issue.

That said. What you want to do in `.zap` files is:
 * Enable the server side of the `Network Provisioning` cluster in `examples/all-clusters-app/all-clusters-common/all-clusters-app.zap`
 * Enable the client side of the `Network Provisioning cluster` in `examples/chip-tool/chip-tool.zap`
 *  Enable the client side of the `Network Provisioning cluster` in `src/darwin/Framework/CHIP/chip-tool.zap`

So this way you will have 2 client applications and one server application.

Comment (erjiaqing) [reply src/app/zap-templates/zcl/chip-devices.xml:189]: Done

File: src/app/zap-templates/partials/header.zapt [hunk 8]
```diff
@@ -1,6 +1,6 @@
 /*
  *
- *    Copyright (c) 2021 Project CHIP Authors
+ *    Copyright (c) 2020-2021 Project CHIP Authors
```
Comment (vivien-apple) [inline src/app/zap-templates/partials/header.zapt:3]: I really wish this was done in a different PR to not pollute the diff of this one. It has introduced a bunch of unrelated changes :(

File: src/app/zap-templates/templates/chip/CHIPClustersObjc-src.zapt [hunk 9]
```diff
@@ -358,7 +358,11 @@ private:
         return NO;
     }
 
+<<<<<<< HEAD
+    CHIP_ERROR err = self.cppCluster.{{asCamelCased name false}}(callback{{#chip_server_cluster_command_arguments}}, {{asCamelCased label}}{{#if (isByteString type)}}, {{asCamelCased label}}Len{{/if}}{{/chip_server_cluster_command_arguments}});
+=======
```
Comment (vivien-apple) [inline src/app/zap-templates/templates/chip/CHIPClustersObjc-src.zapt:363]: I guess that's the root of what looks like a merge issue in all the `CHIPClustersObjc.mm` files.

File: src/app/zap-templates/zcl/provisioning.xml [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+        <item name="Label9" value="0xc"/>
+        <item name="Label10" value="0xd"/>
+        <item name="Label11" value="0xe"/>
+        <item name="Label12" value="0xf"/>
+        <item name="Label13" value="0x10"/>
+        <item name="Label14" value="0x11"/>
+        <item name="Label15" value="0x12"/>
+        <item name="UnknownError" value="0x13"/>
+    </enum>
+    <bitmap name="ShadeClosureStatus" type="BITMAP8">
```
Comment (vivien-apple) [inline src/app/zap-templates/zcl/provisioning.xml:41]: `ShadeClosureStatus` really ? :)

File: examples/all-clusters-app/all-clusters-common/all-clusters-app.zap [hunk 11]
```diff
@@ -5633,6 +5593,196 @@
             }
           ]
         },
+        {
+          "name": "Network Provisioning",
+          "code": 43690,
+          "mfgCode": null,
+          "define": "NETWORK_PROVISIONING_CLUSTER",
+          "side": "client",
```
Comment (vivien-apple) [inline examples/all-clusters-app/all-clusters-common/all-clusters-app.zap:5601]: Please disable the `client` side here for the `all-clusters-app`. It has been enabled in `chip-tool` already.

File: examples/all-clusters-app/all-clusters-common/all-clusters-app.zap [hunk 12]
```diff
@@ -5633,6 +5593,196 @@
...
+            }
+          ]
+        },
+        {
+          "name": "Network Provisioning",
+          "code": 43690,
+          "mfgCode": null,
+          "define": "NETWORK_PROVISIONING_CLUSTER",
+          "side": "server",
+          "enabled": 1,
```
Comment (vivien-apple) [inline examples/all-clusters-app/all-clusters-common/all-clusters-app.zap:5701]: I was looking for the code of the server that can handle those commands and I realised the `gen/` folder of the `all-clusters-app` has not been updated despite the fact that you are enabling some clusters on this file.

Looking at your initial comment for the PR I found this sentence: "Support on end device side will be a seperate PR later.".

Enabling some clusters in this file without providing the associated `gen/` changes will basically break the build for others. I think that's because you have not provided the associated `src/app/clusters` code and so the `gen/` folder will not link without it. Maybe you can just provide a `src/app/clusters/network-provisioning` skeleton so it will make the linker happy and you can provide the implementation later ?

**Review by vivien-apple — CHANGES_REQUESTED** [review]

File: examples/all-clusters-app/all-clusters-common/all-clusters-app.zap (hunk 11 above)
Comment (erjiaqing) [reply examples/all-clusters-app/all-clusters-common/all-clusters-app.zap:5601]: :-(
https://github.com/project-chip/connectedhomeip/blob/3e6f2d43b0609f4aa7e3ad7a361781483016f4a5/scripts/tools/zap_generate_chip.sh#L21

Comment (vivien-apple) [reply examples/all-clusters-app/all-clusters-common/all-clusters-app.zap:5601]: Yes. As I said in your IM patch I'm trying to get rid of `chip-templates.json` in order to let applications uses their own `.zap`file.

So that means the python device controller should have its own `.zap` file and can have custom scripts to generate the `CHIPDeviceController-*` bits.

It also means that I need to move `CHIPClusters*-zap` files to be application specific. So applications will just generate the right set of methods based on what is in their `.zap` file.

I also project to move some of the content of `src/app/encoder.cpp` and `src/app/chip-zcl-zpro-codec-api.h` inside `src/controller/CHIPClusters*` (which would be `gen/CHIPClusters*`. files). And mark those files as deprecated since they can't be removed until all the consumers stop using them. 
For this part I somehow hope to find a way to switch to the new IM instead of doing that since it just seems temporary work...

Finally we would need to find a way to make `CHIPClusters*` not glued to the controller CHIPDevice, maybe to a base class that `src/controller/CHIPDevice` inherits from,  so it can be used from the accessory side as well.

Comment (erjiaqing) [reply examples/all-clusters-app/all-clusters-common/all-clusters-app.zap]: I copied `all-clusters-app.zap` to `src/controller/controller-clusters.zap`, which makes sense for me.

File: examples/all-clusters-app/all-clusters-common/all-clusters-app.zap (hunk 12 above)
Comment (erjiaqing) [reply examples/all-clusters-app/all-clusters-common/all-clusters-app.zap]: Done, network provisioning cluster will be added back to all cluster apps after basic network commissioning handlers are finished.

**Comment by erjiaqing** [comment]
@vivien-apple All comment resolved, PTAL.

**Review by msandstedt — APPROVED** [review]

**Review by vivien-apple — APPROVED** [review]

**Comment by erjiaqing** [comment]
@woody-apple PTAL
BTW, the pullapprove says "CLA must be agreed to by all contributors", but why?

**Review by jelderton — APPROVED** [review]
